torch/testing/_internal/custom_op_db.py

```diff
@@ -56,6 +66,22 @@ def numpy_mul_abstract(x, y):
     assert x.device == y.device
     return (x * y).contiguous()
 
+@numpy_mul.impl_save_for_backward()
+def numpy_mul_save_for_backward(inputs, output):
+    saved = {}
+    saved['x_requires_grad'] = inputs.x.requires_grad
+    saved['y_requires_grad'] = inputs.y.requires_grad
+    # Optimization: only save what is necessary
+    saved['y'] = inputs.y if inputs.x.requires_grad else None
+    saved['x'] = inputs.x if inputs.y.requires_grad else None
+    return saved
+
+@numpy_mul.impl_backward()
+def numpy_mul_backward(ctx, saved, grad_out):
+    grad_x = grad_out * saved['y'] if saved['x_requires_grad'] else None
```

Comment:
The ctx should be telling you what to compute gradients for really.
This is fine as a test but we should not recommend users do this. We should just expose `needs_input_grad` on the ctx.

Reply:
Agree, I didn't add needs_input_grad in this PR because it was getting long.

Reply:
I can add it as a follow-up, unless you want to see it in this PR?

Reply:
No need to add it here as long as this is not in any user-facing example!